sn_networking/src/external_address.rs

```diff
@@ -0,0 +1,261 @@
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+#![allow(dead_code)]
+
+use crate::{driver::NodeBehaviour, mulitaddr_get_ip};
+use itertools::Itertools;
+use libp2p::{Multiaddr, Swarm};
+use std::{collections::HashMap, net::IpAddr};
+
+const MAX_REPORTS_BEFORE_CONFIRMATION: u8 = 3;
+const MAX_CANDIDATES: usize = 50;
+
+/// Manages the external addresses of a Public node. For a relayed node, the RelayManager should deal with
+/// adding and removing external addresses.
+// TODO:
+// 1. if the max candidate is reached, kick out the oldest candidate sorted by # of reports
+#[derive(Debug)]
+pub struct ExternalAddressManager {
+    /// All the external addresses of the node
+    address_states: Vec<ExternalAddressState>,
+    current_ip_address: Option<IpAddr>,
+
+    /// The maximum number of reports before an candidate address is confirmed
+    pub max_reports_before_confirmation: u8,
+    /// The maximum number of candidates to store
+    pub max_candidates: usize,
+}
+
+impl ExternalAddressManager {
+    pub fn new() -> Self {
+        Self {
+            address_states: Vec::new(),
+            current_ip_address: None,
+            max_reports_before_confirmation: MAX_REPORTS_BEFORE_CONFIRMATION,
+            max_candidates: MAX_CANDIDATES,
+        }
+    }
+
+    pub fn set_reports_before_confirmation(&mut self, max_reports_before_confirmation: u8) {
+        self.max_reports_before_confirmation = max_reports_before_confirmation;
+    }
+
+    pub fn set_max_candidates(&mut self, max_candidates: usize) {
+        self.max_candidates = max_candidates;
+    }
+
+    /// Get the list of candidate addresses
+    pub fn get_candidate_addresses(&self) -> Vec<&Multiaddr> {
+        self.address_states
+            .iter()
+            .filter_map(|state| {
+                if let ExternalAddressState::Candidate { address, .. } = state {
+                    Some(address)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    /// Get the list of confirmed addresses
+    pub fn get_confirmed_addresses(&self) -> Vec<&Multiaddr> {
+        self.address_states
+            .iter()
+            .filter_map(|state| {
+                if let ExternalAddressState::Confirmed { address, .. } = state {
+                    Some(address)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    pub fn add_external_address(&mut self, address: Multiaddr, swarm: &mut Swarm<NodeBehaviour>) {
+        debug!("Adding external address to manager: {address:?}");
+        // TODO: have a filter to filter out ill formed addresses.
```

Comment:
I think it would be good to have a normalize function, that strips the `p2p`? I think I've seen mostly two forms of multiaddr, one with `p2p` and without.

At least, it is assumed we're not talking about `p2p-circuit` addresses here, right?

Reply:
Yeah, I've made the modification to do just that. `craft_external_address`.

And yes, we don't deal with `p2p-circuit` addresses here, those are managed by `relay manager`.